nomad/service_registration_endpoint.go

```diff
@@ -0,0 +1,410 @@
+package nomad
+
+import (
+	"time"
+
+	"github.com/armon/go-metrics"
+	"github.com/hashicorp/go-memdb"
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/nomad/acl"
+	"github.com/hashicorp/nomad/nomad/state"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// ServiceRegistration encapsulates the service registrations RPC endpoint
+// which is callable via the ServiceRegistration RPCs and externally via the
+// "/v1/service{s}" HTTP API.
+type ServiceRegistration struct {
+	srv *Server
+
+	// ctx provides context regarding the underlying connection, so we can
+	// perform TLS certificate validation on internal only endpoints.
+	ctx *RPCContext
+}
+
+// Upsert creates or updates service registrations held within Nomad. This RPC
+// is only callable by Nomad nodes.
+func (s *ServiceRegistration) Upsert(
+	args *structs.ServiceRegistrationUpsertRequest,
+	reply *structs.ServiceRegistrationUpsertResponse) error {
+
+	// Ensure the connection was initiated by a client if TLS is used.
+	if err := validateTLSCertificateLevel(s.srv, s.ctx, tlsCertificateLevelClient); err != nil {
+		return err
+	}
+
+	if done, err := s.srv.forward(structs.ServiceRegistrationUpsertRPCMethod, args, args, reply); done {
+		return err
+	}
```

Comment:
If we check that the connection is from a client, won't it then fail once we forward the RPC and pass through that same check on the leader? This endpoint checks the secret ID, so I don't know that we need to manually check the TLS as well (other than the connection check we already have).

Note that we're not making this check on the delete RPCs because they can be end-user initiated, but then we're checking the ACL _or_ node secret.

Reply:
`validateTLSCertificateLevel` is hierarchical so server certificates are valid whereever client certificates are valid.

However now I'm not remembering why we ever need to check `tlsCertificateLevelClient` because `rpc.go` already requires a `client.*.nomad` or `server.*.nomad` certificate: https://github.com/hashicorp/nomad/blob/v1.2.6/nomad/server.go#L541-L557

@lgfa29 Can you refresh my memory why we need `validateTLSCertificateLevel(..., tlsCertificateLevelClient)`? (The equivalent for `tlsCertificateLevelServer` makes sense to keep clients out of server-only RPCs.)

For @jrasell and this PR: this code will work as intended and not break in the face of RPC forwarding. We can leave it alone and if the rest of us realize it's not required we can remove it along with the other `tlsCertificateLevelClient` checks.

Reply:
I think it's poor naming on my part. `tlsCertificateLevelClient` will check for a client cert and, if that fails, check for server:
https://github.com/hashicorp/nomad/blob/main/nomad/util.go#L301-L306

I'm not sure why the check in `rpc.go` is not enough though. Without these checks my [tests](https://github.com/hashicorp/nomad/blob/main/nomad/rpc_test.go#L1104) failed. Maybe that part of code is not exercised in the test?